text/000-clean-up-Reals.md

```diff
@@ -0,0 +1,133 @@
+
+# Detailed design
+
+The detailed design is to be discussed with Reals maintainers (namely Hugo
+Herbelin, Laurent Théry and Guillaume Melquiond) and/or other Coq developers.
+
+Here is a first rough and incomplete plan:
+1. Yet another PR about `RIneq` completing results about division and
+   subtraction and adding common one-step reasoning lemmas, for instance
+   "changing the side of a term" in a equation or an inequality.
+   These are needed to obtain smaller proofs afterwards and correspond to usual
+   mathematical practice for, e.g. high school students.
+2. Incorporate all parts of `Rfunctions` inside `Rfunctions` itself in, say, the
+   same order as they are required. This could be split in separate PR (for
+   instance one for `Rsqr`, one for `Rmin` and `Rmax`, etc). The continuous
+   integration would tell if the small files need to be kept with a deprecation
+   period or can be thrown away directly.
+3. The same method can then be applied to the other main parts of the library,
+   namely `SeqSeries` `Rtrigo` `Ranalysis` and `RiemannInt`.
+
+The following [branch](https://github.com/Villetaneuse/coq/tree/Reals_prototype)
+contains an incomplete prototype with `Rfunctions` and `SeqSeries` as single
+files. While it still needs more polishing and requires advice, it shows that it
+is possible to considerably reduce the size of proofs, without any automation
+except that of the `easy` tactic.
+
+We do not, at this stage, wish to change the underlying logic of the library. We
+can work with this middle ground between intuitionistic and classical logic.
+While we aim at making important changes, we also wish to preserve, to some
+extend, backward compatibility.
+
+# Drawbacks
+
+- The major drawback is probably that it may be time consuming for the
+  maintainers. It could be deemed not worth it.
+- Giving alternative (better) definitions could lead to increase unreasonably
+  the number of lemmas, this will have to be carefully balanced.
+- Having less files *could* lead, in a heavily multithreaded environment to
+  increased compilation time (but we deem it unlikely).
+
+# Alternatives
+
+- It may be possible to be less ambitious regarding the changes, only cleaning
+  up some proofs and deprecating some lemmas, without touching the general
+  structure of the library.
+- Another possible design was described by Vincent Semeria: have a first layer
+  of constructive mathematics and a second layer specialized using stronger
+  axioms. It would be beautiful but it does not solve our issues.
+- The library `mathcomp-analysis` is a very advanced analysis library built on
+  top of Mathcomp, ssreflect and Hierarchy Builder.
+  A definition of the real numbers is given (taken, if we're not mistaken from
+  the pre-existing library coq-alternate-reals) but is seldom, if ever, used.
+  Its lemmas are very general and do not rely on a specific construction of the
+  real numbers, but are parameterized by instances of type `realType`.
```

Comment:
The preexisting library `coq-alternate-reals provided` a structure `realType` for a real number object in type theory, and we indeed use it, rather than a specific implementation. There is an instanciation which uses Coq reals implementation: the famous [`Rstruct` file](https://github.com/math-comp/analysis/blob/master/theories/Rstruct.v)).

Indeed the `realType` structure is used throughout the library, but that does not make the theory logically more general, since all `realType` objects should be isomorphic (the proof is not provided). The main reason is that we may want to apply theorems to both `R` and `{x : C | x \is a Num.real}` or `{x : \bar R | x \is a fin_num}`, but this happens to be not so essential and we might revert back to using an arbitrary implementation of the reals.

Reply:
I must admit I did not have enough time to browse the `math-comp/analysis` as it deserves. Thank you very much for these corrections.

Reply:
I have corrected the statements about math-comp and added a reference to math-classes and coq/coq#17877.